File: apps/web/src/lib/enterprise.ts
import { clearSession, getApiUrl, loadSession, refresh, saveSession } from "@/lib/api";

export type EnterpriseOrg = {
  id: string;
  name: string;
  slug: string;
  franchise: boolean;
  created_at: string | null;
};

export type EnterpriseLocation = {
  id: string;
  organization_id: string;
  shop_id: string;
  name: string;
  code: string;
  region: string | null;
  timezone: string;
  active: boolean;
};

export type CentralDashboard = {
  organization_id: string;
  organization_name: string;
  generated_at: string;
  location_count: number;
  kpis: { id: string; label: string; value: number; unit: string }[];
  locations: {
    location_id: string;
    location_name: string;
    code: string;
    revenue: number;
    appointments: number;
    ai_success_rate: number;
    retention: number;
    customers: number;
  }[];
  brand: Record<string, unknown>;
  policy_count: number;
  audit_recent: number;
  sso_enabled: boolean;
};

export type AiPolicy = {
  id: string;
  name: string;
  effect: string;
  scope: string;
  rules: Record<string, unknown>;
  priority: number;
  enabled: boolean;
  location_id: string | null;
};

export type AuditRow = {
  id: string;
  action: string;
  resource: string;
  resource_id: string | null;
  actor_email: string | null;
  details: Record<string, unknown>;
  created_at: string | null;
};

export type GatewayRoute = {
  id: string;
  path_prefix: string;
  upstream: string;
  auth: string;
  required_role: string | null;
  rate_limit_rpm: number;
  enabled: boolean;
  description: string;
};

async function authFetch(path: string, init?: RequestInit): Promise<Response> {
  const session = loadSession();
  if (!session) throw new Error("Not signed in");
  const url = `${getApiUrl()}${path}`;
  const headers: Record<string, string> = {
    Authorization: `Bearer ${session.accessToken}`,
    ...(init?.body ? { "Content-Type": "application/json" } : {}),
    ...(init?.headers as Record<string, string> | undefined),
  };
  let res = await fetch(url, { ...init, headers });
  if (res.status === 401) {
    const next = await refresh(session.refreshToken);
    if (!next) {
      clearSession();
      throw new Error("Session expired");
    }
    saveSession(next);
    headers.Authorization = `Bearer ${next.accessToken}`;
    res = await fetch(url, { ...init, headers });
  }
  return res;
}

async function json<T>(res: Response): Promise<T> {
  if (!res.ok) {
    const body = await res.json().catch(() => ({}));
    throw new Error(body.detail || body.message || `Request failed (${res.status})`);
  }
  return res.json() as Promise<T>;
}

export async function listOrgs(): Promise<EnterpriseOrg[]> {
  return json(await authFetch("/v1/enterprise/organizations"));
}

export async function seedEnterpriseOrg(): Promise<EnterpriseOrg> {
  return json(await authFetch("/v1/enterprise/organizations/seed", { method: "POST", body: "{}" }));
}

export async function getCentralDashboard(orgId: string): Promise<CentralDashboard> {
  return json(await authFetch(`/v1/enterprise/organizations/${orgId}/dashboard`));
}

export async function listEnterpriseLocations(orgId: string): Promise<EnterpriseLocation[]> {
  return json(await authFetch(`/v1/enterprise/organizations/${orgId}/locations`));
}

export async function listPolicies(orgId: string): Promise<AiPolicy[]> {
  return json(await authFetch(`/v1/enterprise/organizations/${orgId}/policies`));
}

export async function listAudit(orgId: string): Promise<AuditRow[]> {
  return json(await authFetch(`/v1/enterprise/organizations/${orgId}/audit?limit=40`));
}

export async function getBrand(orgId: string): Promise<Record<string, unknown>> {
  return json(await authFetch(`/v1/enterprise/organizations/${orgId}/brand`));
}

export async function updateBrand(orgId: string, body: Record<string, unknown>): Promise<Record<string, unknown>> {
  return json(
    await authFetch(`/v1/enterprise/organizations/${orgId}/brand`, {
      method: "PUT",
      body: JSON.stringify(body),
    }),
  );
}

export async function getFranchiseAnalytics(orgId: string): Promise<Record<string, unknown>> {
  return json(await authFetch(`/v1/enterprise/organizations/${orgId}/franchise-analytics`));
}

export async function listGatewayRoutes(): Promise<GatewayRoute[]> {
  return json(await authFetch("/v1/enterprise/gateway/routes"));
}

export async function listRoleHierarchy(): Promise<{ roles: { role: string; rank: number; label: string }[] }> {
  return json(await authFetch("/v1/enterprise/roles"));
}

export async function beginSso(orgId: string, email?: string): Promise<Record<string, unknown>> {
  return json(
    await authFetch(`/v1/enterprise/organizations/${orgId}/sso/begin`, {
      method: "POST",
      body: JSON.stringify({ email }),
    }),
  );
}

export async function evaluatePolicy(
  orgId: string,
  body: { intent?: string; channel?: string },
): Promise<Record<string, unknown>> {
  return json(
    await authFetch(`/v1/enterprise/organizations/${orgId}/policies/evaluate`, {
      method: "POST",
      body: JSON.stringify(body),
    }),
  );
}

export async function createOrg(body: {
  name: string;
  slug: string;
  franchise?: boolean;
}): Promise<EnterpriseOrg> {
  return json(
    await authFetch("/v1/enterprise/organizations", {
      method: "POST",
      body: JSON.stringify({ franchise: true, ...body }),
    }),
  );
}

export async function addLocation(
  orgId: string,
  body: { shop_id: string; name: string; code: string; region?: string; timezone?: string },
): Promise<EnterpriseLocation> {
  return json(
    await authFetch(`/v1/enterprise/organizations/${orgId}/locations`, {
      method: "POST",
      body: JSON.stringify(body),
    }),
  );
}

export async function listMemberships(orgId: string): Promise<
  { id: string; user_id: string; email: string; role: string; location_ids: string[] }[]
> {
  return json(await authFetch(`/v1/enterprise/organizations/${orgId}/memberships`));
}

export async function grantMembership(
  orgId: string,
  body: { user_id: string; email: string; role: string; location_ids?: string[] },
): Promise<Record<string, unknown>> {
  return json(
    await authFetch(`/v1/enterprise/organizations/${orgId}/memberships`, {
      method: "POST",
      body: JSON.stringify(body),
    }),
  );
}

export async function createPolicy(
  orgId: string,
  body: {
    name: string;
    effect: string;
    scope?: string;
    rules?: Record<string, unknown>;
    priority?: number;
    enabled?: boolean;
    location_id?: string;
  },
): Promise<AiPolicy> {
  return json(
    await authFetch(`/v1/enterprise/organizations/${orgId}/policies`, {
      method: "POST",
      body: JSON.stringify(body),
    }),
  );
}

export async function createApiKey(
  orgId: string,
  body: { name: string; scopes?: string[]; rate_limit_rpm?: number },
): Promise<{
  id: string;
  name: string;
  key_prefix: string;
  api_key: string;
  scopes: string[];
  rate_limit_rpm: number;
}> {
  return json(
    await authFetch(`/v1/enterprise/organizations/${orgId}/api-keys`, {
      method: "POST",
      body: JSON.stringify(body),
    }),
  );
}

export async function configureSso(
  orgId: string,
  body: {
    provider: string;
    client_id: string;
    issuer_url: string;
    domains?: string[];
    role_mapping?: Record<string, string>;
    enabled?: boolean;
    metadata_url?: string;
    client_secret?: string;
    redirect_uri?: string;
    require_sso?: boolean;
  },
): Promise<Record<string, unknown>> {
  return json(
    await authFetch(`/v1/enterprise/organizations/${orgId}/sso`, {
      method: "PUT",
      body: JSON.stringify(body),
    }),
  );
}

export async function completeSso(
  orgId: string,
  body: { state: string; email?: string; code?: string; external_role?: string },
): Promise<Record<string, unknown>> {
  return json(
    await authFetch(`/v1/enterprise/organizations/${orgId}/sso/complete`, {
      method: "POST",
      body: JSON.stringify(body),
    }),
  );
}

async function publicJson(res: Response): Promise<Record<string, unknown>> {
  if (!res.ok) {
    const text = await res.text();
    let detail = text;
    try {
      const parsed = JSON.parse(text) as { detail?: string };
      if (typeof parsed.detail === "string") detail = parsed.detail;
    } catch {
      /* keep text */
    }
    throw new Error(detail || res.statusText);
  }
  return res.json() as Promise<Record<string, unknown>>;
}

/** Public IdP callback — no Bearer token. */
export async function completeSsoCallback(body: {
  state: string;
  email?: string;
  code?: string;
  external_role?: string;
}): Promise<Record<string, unknown>> {
  return publicJson(
    await fetch(`${getApiUrl()}/v1/enterprise/sso/callback`, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(body),
    }),
  );
}

export async function getSsoStatus(orgSlug: string): Promise<{
  organization_id: string;
  organization_slug: string;
  sso_enabled: boolean;
  require_sso: boolean;
  provider: string | null;
}> {
  const res = await fetch(
    `${getApiUrl()}/v1/enterprise/sso/status?org_slug=${encodeURIComponent(orgSlug.trim())}`,
  );
  return publicJson(res) as Promise<{
    organization_id: string;
    organization_slug: string;
    sso_enabled: boolean;
    require_sso: boolean;
    provider: string | null;
  }>;
}

export async function beginSsoPublic(body: {
  org_slug: string;
  email?: string;
}): Promise<Record<string, unknown>> {
  return publicJson(
    await fetch(`${getApiUrl()}/v1/enterprise/sso/begin`, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(body),
    }),
  );
}

export async function authorizeGateway(body: {
  path: string;
  api_key?: string;
  role?: string;
  client_id?: string;
}): Promise<Record<string, unknown>> {
  return json(
    await authFetch("/v1/enterprise/gateway/authorize", {
      method: "POST",
      body: JSON.stringify(body),
    }),
  );
}

export async function getEnterpriseMetrics(): Promise<Record<string, unknown>> {
  return json(await authFetch("/v1/enterprise/metrics/summary"));
}

File: apps/api/app/api/routers/enterprise.py
"""Enterprise HTTP API — multi-location, SSO, policies, gateway, central dashboard."""

from __future__ import annotations

from datetime import datetime
from typing import Any
from uuid import UUID

from fastapi import APIRouter, Depends, HTTPException, Query
from pydantic import BaseModel, Field

from app.api.deps import CurrentUser, get_current_user
from app.enterprise.enums import EnterpriseRole, PolicyEffect, PolicyScope, SsoProvider
from app.enterprise.factory import EnterpriseRuntime, get_enterprise_runtime
from app.enterprise.gateway import GatewayDenied, RateLimited

router = APIRouter(prefix="/v1/enterprise", tags=["enterprise"])


def _rt() -> EnterpriseRuntime:
    return get_enterprise_runtime()


class OrgCreate(BaseModel):
    name: str
    slug: str
    franchise: bool = True


class LocationCreate(BaseModel):
    shop_id: UUID
    name: str
    code: str
    region: str | None = None
    timezone: str = "America/Los_Angeles"


class MembershipGrant(BaseModel):
    user_id: UUID
    email: str
    role: EnterpriseRole
    location_ids: list[UUID] = Field(default_factory=list)


class BrandUpdate(BaseModel):
    product_name: str | None = None
    primary_color: str | None = None
    accent_color: str | None = None
    logo_url: str | None = None
    favicon_url: str | None = None
    support_email: str | None = None
    custom_domain: str | None = None
    login_tagline: str | None = None
    hide_powered_by: bool | None = None


class PolicyCreate(BaseModel):
    name: str
    effect: PolicyEffect
    scope: PolicyScope = PolicyScope.ORGANIZATION
    rules: dict[str, Any] = Field(default_factory=dict)
    location_id: UUID | None = None
    priority: int = 100
    enabled: bool = True


class PolicyEval(BaseModel):
    intent: str | None = None
    channel: str | None = None
    location_id: UUID | None = None
    context: dict[str, Any] = Field(default_factory=dict)


class SsoConfigure(BaseModel):
    provider: SsoProvider
    client_id: str
    issuer_url: str
    domains: list[str] = Field(default_factory=list)
    role_mapping: dict[str, str] = Field(default_factory=dict)
    enabled: bool = True
    metadata_url: str | None = None
    client_secret: str | None = None
    redirect_uri: str | None = None
    require_sso: bool = False


class SsoBegin(BaseModel):
    email: str | None = None


class SsoBeginPublic(BaseModel):
    org_slug: str = Field(min_length=1, max_length=128)
    email: str | None = None


class SsoComplete(BaseModel):
    state: str
    email: str | None = None
    code: str | None = None
    external_role: str | None = None


class ApiKeyCreate(BaseModel):
    name: str
    scopes: list[str] = Field(default_factory=lambda: ["enterprise:read"])
    rate_limit_rpm: int = 120


class GatewayAuthorize(BaseModel):
    path: str
    api_key: str | None = None
    role: EnterpriseRole | None = None
    client_id: str = "web"


class OrgOut(BaseModel):
    id: UUID
    name: str
    slug: str
    franchise: bool
    created_at: datetime | None = None


class LocationOut(BaseModel):
    id: UUID
    organization_id: UUID
    shop_id: UUID
    name: str
    code: str
    region: str | None
    timezone: str
    active: bool


def _brand_out(brand) -> dict[str, Any]:
    return {
        "organization_id": str(brand.organization_id),
        "product_name": brand.product_name,
        "primary_color": brand.primary_color,
        "accent_color": brand.accent_color,
        "logo_url": brand.logo_url,
        "favicon_url": brand.favicon_url,
        "support_email": brand.support_email,
        "custom_domain": brand.custom_domain,
        "login_tagline": brand.login_tagline,
        "hide_powered_by": brand.hide_powered_by,
        "css_vars": brand.css_vars,
    }

@router.get("/roles")
async def role_hierarchy(user: CurrentUser = Depends(get_current_user), rt: EnterpriseRuntime = Depends(_rt)):
    _ = user
    return {"roles": rt.service.role_hierarchy()}


@router.post("/organizations", response_model=OrgOut)
async def create_org(
    body: OrgCreate,
    user: CurrentUser = Depends(get_current_user),
    rt: EnterpriseRuntime = Depends(_rt),
) -> OrgOut:
    try:
        org = rt.service.create_organization(
            name=body.name,
            slug=body.slug,
            franchise=body.franchise,
            owner_user_id=user.user_id,
            owner_email=user.email or "owner@example.com",
        )
    except ValueError as exc:
        raise HTTPException(status_code=400, detail=str(exc)) from exc
    return OrgOut(id=org.id, name=org.name, slug=org.slug, franchise=org.franchise, created_at=org.created_at)


@router.post("/organizations/seed", response_model=OrgOut)
async def seed_org(
    user: CurrentUser = Depends(get_current_user),
    rt: EnterpriseRuntime = Depends(_rt),
) -> OrgOut:
    org = rt.service.seed_demo(
        owner_user_id=user.user_id,
        owner_email=user.email or "owner@example.com",
        primary_shop_id=user.shop_id,
    )
    return OrgOut(id=org.id, name=org.name, slug=org.slug, franchise=org.franchise, created_at=org.created_at)


@router.get("/organizations", response_model=list[OrgOut])
async def list_orgs(user: CurrentUser = Depends(get_current_user), rt: EnterpriseRuntime = Depends(_rt)):
    _ = user
    return [
        OrgOut(id=o.id, name=o.name, slug=o.slug, franchise=o.franchise, created_at=o.created_at)
        for o in rt.service.list_orgs()
    ]


@router.post("/organizations/{org_id}/locations", response_model=LocationOut)
async def add_location(
    org_id: UUID,
    body: LocationCreate,
    user: CurrentUser = Depends(get_current_user),
    rt: EnterpriseRuntime = Depends(_rt),
) -> LocationOut:
    try:
        loc = rt.service.add_location(
            org_id,
            shop_id=body.shop_id,
            name=body.name,
            code=body.code,
            region=body.region,
            timezone=body.timezone,
            actor_user_id=user.user_id,
        )
    except KeyError as exc:
        raise HTTPException(status_code=404, detail=str(exc)) from exc
    return LocationOut(
        id=loc.id,
        organization_id=loc.organization_id,
        shop_id=loc.shop_id,
        name=loc.name,
        code=loc.code,
        region=loc.region,
        timezone=loc.timezone,
        active=loc.active,
    )


@router.get("/organizations/{org_id}/locations", response_model=list[LocationOut])
async def list_locations(
    org_id: UUID,
    user: CurrentUser = Depends(get_current_user),
    rt: EnterpriseRuntime = Depends(_rt),
):
    _ = user
    try:
        locs = rt.service.list_locations(org_id)
    except KeyError as exc:
        raise HTTPException(status_code=404, detail=str(exc)) from exc
    return [
        LocationOut(
            id=l.id,
            organization_id=l.organization_id,
            shop_id=l.shop_id,
            name=l.name,
            code=l.code,
            region=l.region,
            timezone=l.timezone,
            active=l.active,
        )
        for l in locs
    ]


@router.post("/organizations/{org_id}/memberships")
async def grant_membership(
    org_id: UUID,
    body: MembershipGrant,
    user: CurrentUser = Depends(get_current_user),
    rt: EnterpriseRuntime = Depends(_rt),
):
    _ = user
    m = rt.service.grant_membership(
        org_id,
        user_id=body.user_id,
        email=body.email,
        role=body.role,
        location_ids=body.location_ids,
        actor_role=EnterpriseRole.FRANCHISE_OWNER,
    )
    return {
        "id": str(m.id),
        "role": m.role.value,
        "email": m.email,
        "location_ids": [str(x) for x in m.location_ids],
    }


@router.get("/organizations/{org_id}/memberships")
async def list_memberships(
    org_id: UUID,
    user: CurrentUser = Depends(get_current_user),
    rt: EnterpriseRuntime = Depends(_rt),
):
    _ = user
    return [
        {
            "id": str(m.id),
            "user_id": str(m.user_id),
            "email": m.email,
            "role": m.role.value,
            "location_ids": [str(x) for x in m.location_ids],
        }
        for m in rt.service.list_memberships(org_id)
    ]


@router.get("/organizations/{org_id}/dashboard")
async def central_dashboard(
    org_id: UUID,
    user: CurrentUser = Depends(get_current_user),
    rt: EnterpriseRuntime = Depends(_rt),
):
    _ = user
    try:
        dash = rt.service.central_dashboard(org_id)
    except KeyError as exc:
        raise HTTPException(status_code=404, detail=str(exc)) from exc
    return {
        "organization_id": str(dash.organization_id),
        "organization_name": dash.organization_name,
        "generated_at": dash.generated_at.isoformat(),
        "location_count": dash.location_count,
        "kpis": dash.kpis,
        "locations": [
            {
                "location_id": str(l.location_id),
                "location_name": l.location_name,
                "code": l.code,
                "revenue": l.revenue,
                "appointments": l.appointments,
                "ai_success_rate": l.ai_success_rate,
                "retention": l.retention,
                "customers": l.customers,
            }
            for l in dash.locations
        ],
        "brand": dash.brand,
        "policy_count": dash.policy_count,
        "audit_recent": dash.audit_recent,
        "sso_enabled": dash.sso_enabled,
    }


@router.get("/organizations/{org_id}/franchise-analytics")
async def franchise_analytics(
    org_id: UUID,
    user: CurrentUser = Depends(get_current_user),
    rt: EnterpriseRuntime = Depends(_rt),
):
    _ = user
    fa = rt.service.franchise_analytics(org_id)
    return {
        "organization_id": str(fa.organization_id),
        "generated_at": fa.generated_at.isoformat(),
        "totals": fa.totals,
        "rankings": fa.rankings,
        "locations": [
            {
                "location_id": str(l.location_id),
                "location_name": l.location_name,
                "code": l.code,
                "revenue": l.revenue,
                "appointments": l.appointments,
                "ai_success_rate": l.ai_success_rate,
                "retention": l.retention,
                "customers": l.customers,
            }
            for l in fa.locations
        ],
    }


@router.put("/organizations/{org_id}/brand")
async def update_brand(
    org_id: UUID,
    body: BrandUpdate,
    user: CurrentUser = Depends(get_current_user),
    rt: EnterpriseRuntime = Depends(_rt),
):
    _ = user
    brand = rt.service.update_brand(org_id, **body.model_dump(exclude_unset=True))
    return _brand_out(brand)


@router.get("/organizations/{org_id}/brand")
async def get_brand(
    org_id: UUID,
    user: CurrentUser = Depends(get_current_user),
    rt: EnterpriseRuntime = Depends(_rt),
):
    _ = user
    brand = rt.service.get_brand(org_id)
    if brand is None:
        raise HTTPException(status_code=404, detail="Brand not found")
    return _brand_out(brand)


@router.post("/organizations/{org_id}/policies")
async def create_policy(
    org_id: UUID,
    body: PolicyCreate,
    user: CurrentUser = Depends(get_current_user),
    rt: EnterpriseRuntime = Depends(_rt),
):
    _ = user
    p = rt.service.save_policy(
        org_id,
        name=body.name,
        effect=body.effect,
        scope=body.scope,
        rules=body.rules,
        location_id=body.location_id,
        priority=body.priority,
        enabled=body.enabled,
    )
    return {
        "id": str(p.id),
        "name": p.name,
        "effect": p.effect.value,
        "scope": p.scope.value,
        "rules": p.rules,
        "priority": p.priority,
        "enabled": p.enabled,
    }


@router.get("/organizations/{org_id}/policies")
async def list_policies(
    org_id: UUID,
    user: CurrentUser = Depends(get_current_user),
    rt: EnterpriseRuntime = Depends(_rt),
):
    _ = user
    return [
        {
            "id": str(p.id),
            "name": p.name,
            "effect": p.effect.value,
            "scope": p.scope.value,
            "rules": p.rules,
            "priority": p.priority,
            "enabled": p.enabled,
            "location_id": str(p.location_id) if p.location_id else None,
        }
        for p in rt.service.list_policies(org_id)
    ]


@router.post("/organizations/{org_id}/policies/evaluate")
async def evaluate_policy(
    org_id: UUID,
    body: PolicyEval,
    user: CurrentUser = Depends(get_current_user),
    rt: EnterpriseRuntime = Depends(_rt),
):
    _ = user
    return rt.service.evaluate_policy(
        org_id,
        intent=body.intent,
        channel=body.channel,
        location_id=body.location_id,
        context=body.context,
    )


@router.put("/organizations/{org_id}/sso")
async def configure_sso(
    org_id: UUID,
    body: SsoConfigure,
    user: CurrentUser = Depends(get_current_user),
    rt: EnterpriseRuntime = Depends(_rt),
):
    _ = user
    cfg = rt.service.configure_sso(
        org_id,
        provider=body.provider,
        client_id=body.client_id,
        issuer_url=body.issuer_url,
        domains=body.domains,
        role_mapping=body.role_mapping,
        enabled=body.enabled,
        metadata_url=body.metadata_url,
        client_secret=body.client_secret,
        redirect_uri=body.redirect_uri,
        require_sso=body.require_sso,
    )
    return {
        "provider": cfg.provider.value,
        "enabled": cfg.enabled,
        "client_id": cfg.client_id,
        "issuer_url": cfg.issuer_url,
        "domains": cfg.domains,
        "oidc_ready": bool((cfg.metadata or {}).get("client_secret")),
        "require_sso": cfg.require_sso,
    }


@router.post("/organizations/{org_id}/sso/begin")
async def sso_begin(
    org_id: UUID,
    body: SsoBegin,
    user: CurrentUser = Depends(get_current_user),
    rt: EnterpriseRuntime = Depends(_rt),
):
    _ = user
    try:
        return rt.service.sso_begin(org_id, email=body.email)
    except ValueError as exc:
        raise HTTPException(status_code=400, detail=str(exc)) from exc


@router.post("/organizations/{org_id}/sso/complete")
async def sso_complete(
    org_id: UUID,
    body: SsoComplete,
    user: CurrentUser = Depends(get_current_user),
    rt: EnterpriseRuntime = Depends(_rt),
):
    _ = user
    _ = org_id
    try:
        return await rt.service.sso_complete(
            state=body.state,
            email=body.email,
            code=body.code,
            external_role=body.external_role,
            user_id=user.user_id,
        )
    except ValueError as exc:
        raise HTTPException(status_code=400, detail=str(exc)) from exc


@router.post("/sso/callback")
async def sso_callback_public(
    body: SsoComplete,
    rt: EnterpriseRuntime = Depends(_rt),
):
    """IdP redirect target — no app session required (OIDC code or demo state)."""
    try:
        return await rt.service.sso_complete(
            state=body.state,
            email=body.email,
            code=body.code,
            external_role=body.external_role,
            user_id=None,
        )
    except ValueError as exc:
        raise HTTPException(status_code=400, detail=str(exc)) from exc


@router.get("/sso/status")
async def sso_status_public(
    org_slug: str = Query(min_length=1, max_length=128),
    rt: EnterpriseRuntime = Depends(_rt),
):
    """Public probe: whether an organization has SSO enabled (for login UI)."""
    try:
        return rt.service.sso_status_by_slug(org_slug)
    except ValueError as exc:
        raise HTTPException(status_code=404, detail=str(exc)) from exc


@router.post("/sso/begin")
async def sso_begin_public(
    body: SsoBeginPublic,
    rt: EnterpriseRuntime = Depends(_rt),
):
    """Start SSO from login page using organization slug (no app session)."""
    try:
        return rt.service.sso_begin_by_slug(body.org_slug, email=body.email)
    except ValueError as exc:
        raise HTTPException(status_code=400, detail=str(exc)) from exc


@router.get("/organizations/{org_id}/audit")
async def list_audit(
    org_id: UUID,
    limit: int = Query(100, ge=1, le=500),
    user: CurrentUser = Depends(get_current_user),
    rt: EnterpriseRuntime = Depends(_rt),
):
    _ = user
    return [
        {
            "id": str(e.id),
            "action": e.action.value,
            "resource": e.resource,
            "resource_id": e.resource_id,
            "actor_email": e.actor_email,
            "details": e.details,
            "created_at": e.created_at.isoformat() if e.created_at else None,
        }
        for e in rt.service.list_audit(org_id, limit=limit)
    ]


@router.post("/organizations/{org_id}/api-keys")
async def create_api_key(
    org_id: UUID,
    body: ApiKeyCreate,
    user: CurrentUser = Depends(get_current_user),
    rt: EnterpriseRuntime = Depends(_rt),
):
    _ = user
    key, raw = rt.service.create_api_key(
        org_id,
        name=body.name,
        scopes=body.scopes,
        rate_limit_rpm=body.rate_limit_rpm,
    )
    return {
        "id": str(key.id),
        "name": key.name,
        "key_prefix": key.key_prefix,
        "api_key": raw,
        "scopes": key.scopes,
        "rate_limit_rpm": key.rate_limit_rpm,
    }


@router.get("/gateway/routes")
async def gateway_routes(user: CurrentUser = Depends(get_current_user), rt: EnterpriseRuntime = Depends(_rt)):
    _ = user
    return [
        {
            "id": r.id,
            "path_prefix": r.path_prefix,
            "upstream": r.upstream,
            "auth": r.auth.value,
            "required_role": r.required_role.value if r.required_role else None,
            "rate_limit_rpm": r.rate_limit_rpm,
            "enabled": r.enabled,
            "description": r.description,
        }
        for r in rt.gateway.list_routes()
    ]


@router.post("/gateway/authorize")
async def gateway_authorize(
    body: GatewayAuthorize,
    user: CurrentUser = Depends(get_current_user),
    rt: EnterpriseRuntime = Depends(_rt),
):
    try:
        return rt.service.gateway_authorize(
            path=body.path,
            org_id=None,
            api_key=body.api_key,
            role=body.role or EnterpriseRole.ORG_ADMIN,
            client_id=body.client_id or str(user.user_id),
        )
    except GatewayDenied as exc:
        raise HTTPException(status_code=403, detail=str(exc)) from exc
    except RateLimited as exc:
        raise HTTPException(status_code=429, detail=str(exc)) from exc


@router.get("/metrics/summary")
async def metrics(user: CurrentUser = Depends(get_current_user), rt: EnterpriseRuntime = Depends(_rt)):
    _ = user
    return rt.service.metrics()
